xrootd/internal/xrdenc/xrdenc.go

```diff
@@ -0,0 +1,143 @@
+// Copyright 2018 The go-hep Authors.  All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+package xrdenc // import "go-hep.org/x/hep/xrootd/internal/xrdenc"
+
+import "encoding/binary"
+
+// Encoder encodes values to a buffer according to the XRootD protocol.
+type Encoder struct {
+	buf []byte
+}
+
+func (enc *Encoder) Bytes() []byte { return enc.buf }
+
+func (enc *Encoder) WriteU8(v uint8) {
+	enc.buf = append(enc.buf, v)
+}
+
+func (enc *Encoder) WriteU16(v uint16) {
+	var buf [2]byte
+	binary.BigEndian.PutUint16(buf[:], v)
+	enc.buf = append(enc.buf, buf[:]...)
+}
+
+func (enc *Encoder) WriteI32(v int32) {
+	var buf [4]byte
+	binary.BigEndian.PutUint32(buf[:], uint32(v))
+	enc.buf = append(enc.buf, buf[:]...)
+}
+
+func (enc *Encoder) WriteI64(v int64) {
+	var buf [8]byte
+	binary.BigEndian.PutUint64(buf[:], uint64(v))
+	enc.buf = append(enc.buf, buf[:]...)
+}
+
+func (enc *Encoder) WriteBool(v bool) {
+	if v {
+		enc.buf = append(enc.buf, 1)
+	} else {
+		enc.buf = append(enc.buf, 0)
+	}
+}
+
+func (enc *Encoder) WriteLen(n int) {
+	enc.WriteI32(int32(n))
+}
+
+func (enc *Encoder) WriteBytes(vs []byte) {
+	enc.buf = append(enc.buf, vs...)
+}
+
+func (enc *Encoder) WriteStr(str string) {
+	enc.WriteLen(len(str))
+	enc.WriteBytes([]byte(str))
+}
+
+func (enc *Encoder) WriteReserved(n int) {
+	enc.buf = append(enc.buf, make([]byte, n)...)
+}
+
+// Decoder decodes values from a buffer according to the XRootD protocol.
+type Decoder struct {
+	buf []byte
+	pos int
+}
+
+func NewDecoder(data []byte) *Decoder {
+	return &Decoder{buf: data}
+}
+
+func (dec *Decoder) ReadU8() uint8 {
+	o := dec.buf[dec.pos]
+	dec.pos++
+	return o
+}
+
+func (dec *Decoder) Available() int {
```

Comment:
to mimic other Go's stdlib APIs, I would rename this `Len` or `Size`:

```
$> go doc bytes.Buffer.Len
func (b *Buffer) Len() int
    Len returns the number of bytes of the unread portion of the buffer; b.Len()
    == len(b.Bytes()).

$> go doc bytes.Reader.Len
func (r *Reader) Len() int
    Len returns the number of bytes of the unread portion of the slice.

$> go doc bytes.Reader.Size 
func (r *Reader) Size() int64
    Size returns the original length of the underlying byte slice. Size is the
    number of bytes available for reading via ReadAt. The returned value is
    always the same and is not affected by calls to any other method.

```

(so.. perhaps `Len` fits better)

Reply:
Yep, both `Size` and `Available` looked awkward to me and I haven't noticed that `Len` is about unread portion. Thanks!